### DISCUSSION TITLE
Compilation error in AuxKernel.C when new function is added

### DISCUSSION THREAD

**Discussion by abarun22** [idaholab/moose#20792] [Q&A General] [answered]
Hi,
I am working towards implementing the new transfer scheme `(MultiAppVariableVectorSamplePostprocessorTransfer)` as part of building my multi-scale model. Towards this end i wish to add a new function `get_selected_qp` in `RankTwoAux.C` that just return the quadrature point number and this has to be called from within AuxKernel.C. The compiler however does not allow this simple addition  and throws the following error.
![image](https://user-images.githubusercontent.com/40111479/163439036-1526007d-d42c-40f1-a8ca-8ea46170f2b3.png)

This has to be done in the same manner as `computeValue` which is defined and used in `RankTwoAux.h` and `AuxKernel.C` respectively. Unfortunately the addition of new function is not known to the compiler, mainly due to the fact that its templated. The inclusion of `computeValue` suggests that dependencies required from RankTwoAux is already satisfied. I welcome any suggestion/pointers on a possible fix for this error. The related modifications for `get_selected_qp` can be seen here.

```
-------From RankTwoAux.C-----------
template <bool is_ad>
unsigned int RankTwoAuxTempl<is_ad>::get_selected_qp()
{
return _selected_qp;
}
```
<pre><code>
-------From RankTwoAux.h-----------
template <bool is_ad>
class RankTwoAuxTempl : public NodalPatchRecovery
{
public:
  static InputParameters validParams();

  RankTwoAuxTempl(const InputParameters & parameters);

public:
  virtual Real computeValue();
 <strong> virtual unsigned int get_selected_qp(); </strong>

private:
  const GenericMaterialProperty<RankTwoTensor, is_ad> & _tensor;
  const unsigned int _i;
  const unsigned int _j;

  /// whether or not selected_qp has been set
  const bool _has_selected_qp;

  /// The std::vector will be evaluated at this quadpoint only if defined
  const unsigned int _selected_qp;
};
</code></pre>

<code><pre>
--------- From AuxKernel.C -----------------
else /* elemental variables */
  {
    _n_local_dofs = _var.numberOfDofs();
    if (_n_local_dofs == 1) /* p0 */
    {
      ComputeValueType value = 0;
      for (_qp = 0; _qp < _qrule->n_points(); _qp++){
        value += _JxW[_qp] * _coord[_qp] * computeValue();
      }
      value /= (_bnd ? _current_side_volume : _current_elem_volume);        
      if (_var.isFV())
        setDofValueHelper(value);
      else
        // update the variable data referenced by other kernels.
        // Note that this will update the values at the quadrature points too
        // (because this is an Elemental variable)
        _var.setNodalValue(value);
        <strong> printf("get_selected_qp:%d\n",get_selected_qp()); </strong>
</code></pre>

**Comment by GiudGiud** [comment, marked as answer]
So get_selected_qp() is defined in RankTwoAux but you want to use it in AuxKernel?

But RankTwoAux derives from AuxKernel not the other way around, so it must be defined in AuxKernel if it s used there

**Reply by abarun22** [reply]
Yes `get_selected_qp()` is defined in `RankTwoAux` and i want to call this from `AuxKernel.C`. I tried defining this in `AuxKernel.C` through the introduction of header file `"RankTwoAux.h"`. But it did not help. Its quite tricky to call a derived class function inside a parent class. I tried several other options and kind of stuck at this. I wonder how `computeValue()` is called in `AuxKernel.C` without going through all the complicated stuffs. I am doing very similar here.

**Reply by GiudGiud** [reply]
It's because you should not do this. A function needed by a parent class needs to be defined in the parent class, not in the derived class. 

If you REALLY want to do this the other way, do this (pseudo code):
```
if (dynamic_cast<RankTwoAux *>(this))
{
  dynamic_cast<RankTwoAux *>(this)->get_selected_qp();
 }
```

**Reply by abarun22** [reply]
That's an excellent suggestion. It works OK. I wonder how this block of code works in AuxKernel.C without this downcasting procedure
```
for (_qp = 0; _qp < _qrule->n_points(); _qp++){
        value += _JxW[_qp] * _coord[_qp] * computeValue();
}
```
